Lib/Gateway/Paypal.php

```diff
@@ -1,569 +1,1889 @@
+                    throw new \Exception('Empty webhook payload');
+                }
+
+                // Verify signature
+                if ( ! $this->verify_webhook_signature_from_input($raw_input)) {
+                    throw new \Exception('Invalid webhook signature');
+                }
+
+                // Decode and validate
+                $webhook_data = json_decode($raw_input, true);
+                if ( JSON_ERROR_NONE !== json_last_error() ) {
+                    throw new \Exception('Invalid JSON');
+                }
+                if ( ! isset($webhook_data['event_type']) ) {
+                    throw new \Exception('Missing event type');
+                }
 
-            if ( $data['custom']['trial_status'] == 'yes' ) {
-                $paypal_args['p1'] = $data['custom']['trial_duration'];
-                $paypal_args['t1'] = $trial_period;
-                $paypal_args['a1'] = 0;
+                // Log the event type
+                error_log('WPUF PayPal: Received webhook event: ' . $webhook_data['event_type']);
+
+                // Process the webhook
+                $this->process_webhook($raw_input);
+
+                $acknowledged = true;
+            } catch (\Exception $e) {
+                error_log('WPUF PayPal: Webhook error: ' . $e->getMessage());
+                // Optionally, save the failed payload for manual review
             }
-        } else {
-            $paypal_args = [
-                'cmd'           => '_xclick',
-                'business'      => wpuf_get_option( 'paypal_email', 'wpuf_payment' ),
-                'amount'        => $billing_amount,
-                'item_name'     => isset( $data['custom']['post_title'] ) ? $data['custom']['post_title'] : $data['item_name'],
-                'no_shipping'   => '1',
-                'shipping'      => '0',
-                'no_note'       => '1',
-                'currency_code' => $data['currency'],
-                'item_number'   => $data['item_number'],
-                'charset'       => 'UTF-8',
-                'rm'            => '2',
-                'custom'        => wp_json_encode(
-                    [
-                        'type'      => $data['type'],
-                        'user_id'   => $user_id,
-                        'coupon_id' => $coupon_id,
-                        'subtotal'  => $data['subtotal'],
-                        'tax'       => $data['tax'],
-                    ]
-                ),
-                'return'        => $return_url,
-                'notify_url'    => $listener_url,
+
+            // Always acknowledge to PayPal
+            http_response_code(200);
+            echo json_encode(['status' => $acknowledged ? 'ok' : 'error']);
+            exit;
+        }
+    }
+
+    /**
+     * Get current time in UTC
+     */
+    private function get_current_time_utc() {
+        // Get WordPress timezone setting
+        $timezone = wp_timezone();
```

Comment:
please use `wpuf_current_datetime()` and WordPress specific data/time functions. prefare `DateTimeImmutable` class over `DateTime` class